### PR TITLE
Status goals

### PR DESCRIPTION
**Pull request by TihomirovPavel**
### Checklist
- [x] I've run `mvn test` from the root directory to see all new and existing tests pass
- [x] I've followed code style and run and ensured the code style is valid
- [x] I've created new tests if necessary.

### Motivation and Context
#52 

### Description
this functionality is required for tracking goals statuses.

### CODE REVIEW

File: src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java [hunk 1]
```diff
@@ -152,7 +154,6 @@ public void editGoal(@PathParam("id") Long goalId, GoalFormDto newGoalDto) {
     }
 
     @POST
-    @RolesAllowed({"ADMIN", "USER"})
```
Comment (olegvm) [inline src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java]: Why was this removed? Do you want to allow users without these roles to use these methods?

File: src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java [hunk 2]
```diff
@@ -204,5 +203,22 @@ public UserDto getUserById(@PathParam("id") Long id) {
...
     }
+
+    @POST
+    @Path("/mygoals/{id}")
+    public void setStatusAchieved(@PathParam("id") Long goalId) {
+        User user = userStore.getCurrentUser();
+        Optional<Goal> goal = goalStore.getGoalById(goalId);
+        if (goal.isPresent()) {
+            Goal g = goal.get();
+            if (g.getUser() == user) {
```
Comment (olegvm) [inline src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java]: Only primitives, `String`s and `Enum`s may be compared this way. Never use `==` to compare objects unless you are explicitly comparing references to the same object. Use `equals()` (making sure the class implements both `equals()` and `hashCode()`), or compare user IDs.

File: src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java (hunk 1 above)
Comment (TihomirovPavel) [reply src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java]: they were moved to the top of the GoalApi because that all the methods are allowed to all users in GoalApi .

File: src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java (hunk 2 above)
Comment (olegvm) [reply src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java]: It may be possible that they are the same object due to entity persistence framework (Hibernate), however we cannot rely on that.

File: src/main/java/lv/ctco/javaschool/goal/control/GoalStore.java [hunk 3]
```diff
@@ -19,12 +22,26 @@
...
                         "where g.user = :user ", Goal.class)
                 .setParameter("user", user)
                 .getResultList();
+        return CheckStatus(goals);
+    }
+
+    private List<Goal> CheckStatus(List<Goal> goals) {
+        List<Goal> goalsToReturn = new ArrayList<>();
+        for (Goal goal : goals) {
+            if ((goal.getDeadlineDate().isBefore(LocalDate.now())
```
Comment (olegvm) [inline src/main/java/lv/ctco/javaschool/goal/control/GoalStore.java]: Please extract `LocalDate.now()` to a variable and place it outside the `for` loop, to ensure consistent comparison of all goals against a single date and to reduce the amount of calls to `LocalDate.now()`.

File: src/main/webapp/app/js/goal-page.js [hunk 4]
```diff
@@ -135,4 +143,22 @@ function saveEditGoal() {
...
+        "method": "POST",
+        headers: {
+            'Accept': 'application/json',
+            'Content-Type': 'application/json'
+        }
+    }).then(function (response) {
+        if (response.status === 204) {
+            onLoad();
+        } else {
+            console.log(response);
```
Comment (olegvm) [inline src/main/webapp/app/js/goal-page.js]: Might be worth it to extract `console.log()` and `alert()` to a single function which reports the error, and reuse that function over the code.

File: src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java (hunk 1 above)
Comment (olegvm) [reply src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java]: OK. Did not see the class-level annotation at first, due to GitHub collapsing unchanged code.

File: src/main/webapp/app/js/goal-page.js (hunk 4 above)
Comment (TihomirovPavel) [reply src/main/webapp/app/js/goal-page.js]: alert() extracted to function displayErrors() and removed all console.log()

File: src/main/java/lv/ctco/javaschool/goal/control/GoalStore.java (hunk 3 above)
Comment (TihomirovPavel) [reply src/main/java/lv/ctco/javaschool/goal/control/GoalStore.java]: extracted thank you.

File: src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java (hunk 2 above)
Comment (TihomirovPavel) [reply src/main/java/lv/ctco/javaschool/goal/boundary/GoalApi.java]: added equals() and hashCode()

File: src/main/webapp/app/js/goal-page.js [hunk 5]
```diff
@@ -29,17 +29,17 @@ function onLoad() {
...
         }
     }).then(function (response) {
-        if (response.status === 200) {
-            return response.json();
-        } else {
-            alert("Something went wrong! error: "+response.status.toString());
-            return false;
-        }
+        displayError(response, 200);
+        return response.json();
```
Comment (olegvm) [inline src/main/webapp/app/js/goal-page.js]: Will this `return` work as intended in case of error?

File: src/main/webapp/app/js/redirects.js [hunk 6]
```diff
@@ -6,10 +6,10 @@ function logout() {
         });
 }
 function redirectToGoalsAndComments(id) {
-    if (id >= 0) {
-        location.href = path + "/app/goal.jsp?id=" + id;
+    if (id === "-9999") {
```
Comment (olegvm) [inline src/main/webapp/app/js/redirects.js]: What is this `id`?

File: src/main/webapp/app/js/redirects.js [hunk 7]
```diff
@@ -52,4 +52,12 @@ function switchPersonalData() {
     } else {
         block.classList.add("w3-hide");
     }
+}
+
+function displayError(response, expected) {
+    if (response.status === expected) {
```
Comment (olegvm) [inline src/main/webapp/app/js/redirects.js]: The `if` condition should be inverted to avoid an empty block, i.e.:
```
if (response.status !== expected) {
    ...
}
```

File: src/main/webapp/app/js/redirects.js [hunk 8]
```diff
@@ -52,4 +52,12 @@ function switchPersonalData() {
...
+    } else {
+        alert("Something went wrong! error: " + response.status.toString());
+        history.go(-1);
```
Comment (olegvm) [inline src/main/webapp/app/js/redirects.js]: What is the reason for automatically going back to the previous page in this case?

File: src/main/webapp/app/js/goal-page.js (hunk 5 above)
Comment (TihomirovPavel) [reply src/main/webapp/app/js/goal-page.js]: Yes, displayError() does redirect in case of error so the rest of the code will not start.
